### ISSUE TITLE
Provide invert feature for component inputs

### ISSUE THREAD

**Issue by prlaba** [logisim-evolution/logisim-evolution#1510]
Most L-E components use active-high input pins for enabling, selecting, etc.

Oftentimes that polarity is opposite the polarity of the pin found on the corresponding physical components.

For example, the WE and OE pins on L-E's RAM component are active-high, while the corresponding pins on all static RAM ICs are active-low.

While it's easy enough to add inverters to reverse the polarity of signals tied to those pins, it would be better to be able to indicate these pins' polarity as as attribute of the component itself.  Marking a pin active-los would add an active-low "bubble" to the pin, the same way the inverted inputs appear for the AND, OR and other gate components.

This should apply to these components (there may be others):

RAM (WE and OE pins)
ROM (OE pin, if it existed)
D, JK and SR Flip-Flops (Set and Reset pins)
Multiplexer (Enable pin)
Demultiplexer (Enable pin)
Decoder (Enable pin)

(For the Demultiplexer and Decoder components, it would be great to be able to mark their outputs as inverted as well; most Demultiplexer and Decoder ICs have active-low outputs)

Doing this would eliminate the need for inverters to reverse the active-high inputs' polarity.  More importantly, it would allow L-E circuits to more closely reflect the physical components used in those circuits.

**Comment by maehne** [comment]
Pull requests in this direction are welcome!

**Comment by prlaba** [comment]
I wish I were skilled enough w Java to do this.

**Comment by maehne** [comment]
@BFH-ktt1: Would you maybe like to take this up, as you addressed the related issue #1477 with PR #1506?

**Comment by cosineblast** [comment]
With regards to multiplexes and the like, if we did include a circle indicating that the select is negated, given a multiplexer like the one attached in the picture, would it be ok to move the select input downwards by one unit to accommodate the circle?
![shot](https://github.com/logisim-evolution/logisim-evolution/assets/55855728/10ed7f5b-e279-482f-a806-13d157a9e0f9)

**Comment by davidhutchens** [comment]
I am not sure it makes sense to negate the select input to the multiplexer since the select input can be many bits wide and is treated as an integer value. Negating all of them seems strange.

The request was for a negation of the **enable** pin, not the **select** pin.

**Comment by figurantpp** [comment]
This makes sense, thanks for the clarification.

**Comment by maehne** [comment]
> I am not sure it makes sense to negate the select input to the multiplexer since the select input can be many bits wide and is treated as an integer value. Negating all of them seems strange.
> 
> The request was for a negation of the **enable** pin, not the **select** pin.

I agree with @davidhutchens. Nevertheless, just for the record: It is acceptable to offset the pin location when adding the circle, as it is already the case for the basic logic gates, which allow negating their inputs.

**Comment by prlaba** [comment]
In general it would be very useful to be able to invert any enable, set or reset input of a component, to avoid having to use inverters.  This would apply to memories, flip-flops, counters, etc.  Another possible candidate would be the CLK input of any component.

Inverting an input would add a bubble to the displayed component’s input, similar to how inverted inputs appear on the logic gate components.

Paul

**Comment by hewzhew** [comment]
I would be interested in continuing the active-low RAM enable work from #1855 on the current `main`, with clear credit to @cosineblast's original implementation.

Based on the latest maintainer direction in that PR, I propose a focused first slice:

- keep active-high `OE` and `WE` as the defaults, preserving existing `.circ` behavior and the current interface;
- add active-low `OE`/`WE` behavior and the corresponding inversion bubbles for the RAM component;
- mark active-low configurations as not synthesizable rather than increasing the complexity of the RAM HDL generator; and
- add focused simulation, default/serialization, appearance, and HDL-support-gating regression tests.

Should byte/line-enable polarity remain a separate follow-up, or should it be included in this RAM slice? I will keep any implementation local until the intended scope is confirmed.

**Comment by maehne** [comment]
Thanks @hewzhew for volunteering to pick up PR #1855 and bring it to a state allowing it to be merged! The approach you propose sounds good to me. If the activity for all enables are controlled via a single attribute, I personally think that the PR could also considerate byte/line-enable polarity like @davidhutchens suggested.

@BFH-ktt1: Do you agree? Any additional thoughts?
